crates/red_knot_python_semantic/resources/mdtest/unary/instance.md

```diff
@@ -29,3 +31,84 @@ b = NoDunder()
 -b  # error: [unsupported-operator] "Unary operator `-` is unsupported for type `NoDunder`"
 ~b  # error: [unsupported-operator] "Unary operator `~` is unsupported for type `NoDunder`"
 ```
+
+## Not
+
+Not operator is inferred based on
+<https://docs.python.org/3/library/stdtypes.html#truth-value-testing>. All objects are true unless
+the `__bool__` or `__len__` is returning `False`.
+
+[`__len__`](https://docs.python.org/3/reference/datamodel.html#object.__len__) method should return
+an integer and [`__bool__`](https://docs.python.org/3/reference/datamodel.html#object.__bool__)
+method must return a boolean.
+
+```py
+class AlwaysTrue:
+    def __bool__(self) -> Literal[True]:
+        return True
+
+# revealed: Literal[False]
+reveal_type(not AlwaysTrue())
+
+class AlwaysFalse:
+    def __bool__(self) -> Literal[False]:
+        return False
+
+# revealed: Literal[True]
+reveal_type(not AlwaysFalse())
+
+class NoBoolMethod: ...
+
+# revealed: Literal[False]
+reveal_type(not NoBoolMethod())
+
+class LenZero:
+    def __len__(self) -> Literal[0]:
+        return 0
+
+# revealed: Literal[True]
+reveal_type(not LenZero())
+
+class LenNonZero:
+    def __len__(self) -> Literal[1]:
+        return 1
+
+# revealed: Literal[False]
+reveal_type(not LenNonZero())
```

Comment:
Because a subclass could add a `__bool__` method, and the `__bool__` method takes priority over `__len__` for determining boolean value of an instance, I think we cannot conclude anything from the `__len__` method, and we need to just infer `bool` in both of these cases.

If a class has `__len__` but no `__bool__`, we cannot trust `__len__` because a subclass could add a `__bool__`, which would then take precedence. If a class has both `__bool__` and `__len__`, then `__len__` is always irrelevant, `__bool__` determines our answer.

I think this means we can (and must) totally ignore `__len__` when it comes to determining boolean value of a `Type::Instance`, which should simplify the implementation.

(I realize our TODO comment mentioned `__len__`, which just means we didn't think through this carefully enough when we wrote that comment :/ sorry that comment led to extra work.)